docs-internal/multipart.md

```diff
@@ -0,0 +1,328 @@
+# Helidon Multipart Support Proposal
+
+Provide an API for the reactive webserver that enables processing multipart requests and returning multipart responses.
+
+## Proposal
+
+Support 2 models: buffered and reactive.
+
+With the buffered model the request content is fully decoded in memory and all the parts are handed-over at the same time.
+This model enables simple use-cases with small size parts.
+
+With the reactive model the request content is decoded for each chunk and the user can subscribe to the stream of chunks for each part.
+This model enables advanced streaming use-cases with large size parts.
+
+Supporting multipart in the webserver means adding support for a new media type. This is done by adding a new entity type with readers and writers.
+
+Each of the 2 models above will be represented by different entity types: `BufferedMultiPart` and `MultiPart`.
+
+```java
+// buffered model
+request.content().as(BufferedMultiPart.class).thenAccept(/* ... */);
+```
+
+```java
+// reactive model
+request.content().as(MultiPart.class).thenAccept(/* ... */);
+```
+
+## Examples
+
+Let's illustrate with code examples first.
+
+### Enable Multipart Support
+
+The readers and writers for the multipart entity types will be registered by a service, similar to how other media types are supported in the webserver.
+All the examples below assume that the MultiPartSupport is registered as follow:
+
+```java
+Routing.builder()
+    .register(MultiPartSupport.create())
+    // user defined handlers...
+    .build());
+```
+
+### Process a request with the buffered model
+
+The following example consumes each part content as a `JsonObject`:
+
+```java
+request.content().as(BufferedMultiPart.class).thenAccept(mp -> {
+    for(BodyPart bodyPart : mp.bodyParts()){
+        JsonObject json = bodyPart.content().as(JsonObject.class);
+        System.out.println("File uploaded: " + bodyPart.headers().filename());
+        System.out.println(json.toString());
+    }
+    response.send("Files uploaded successful");
+});
+```
+
+### Return a response with the buffered model
+
+The following example echoes the multipart request into the response:
+
+```java
+request.content().as(BufferedMultiPart.class).thenAccept(mp -> {
+    response.send(mp);
+});
+```
+
+The following example returns a multipart response from a list of `JsonObject`:
+
+```java
+BufferedMultiPart mp = BufferedMultiPart.create(
+    Json.createObjectBuilder().add("foo", "bar").build(),
+    Json.createObjectBuilder().add("alice", "bob").build());
+response.send(mp);
+```
```

Comment:
In this case the media-type for each part will be `application/json`, but in general we may need to support the creation from `BodyPart`'s?

Reply:
```java
BufferedMultiPart.create(T ... entities)
``` 
is a shortcut for 

```java
BufferedMultiPart.builder().bodyPart(BodyPart.create(T entity)).build()
```

With the current proposal one use the BodyPart builder to customize each part. See the example named "Return a response with the reactive model".